flink/v1.15/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkParquetReader.java

```diff
@@ -81,26 +75,87 @@ public void testTwoLevelList() throws IOException {
-    writer.close();
+    try (ParquetWriter<GenericRecord> writer =
+        AvroParquetWriter.<GenericRecord>builder(new Path(testFile.toURI()))
+            .withDataModel(GenericData.get())
+            .withSchema(avroSchema)
+            .config("parquet.avro.add-list-element-records", "true")
+            .config("parquet.avro.write-old-list-structure", "true")
+            .build()) {
+      writer.write(expectedRecord);
+    }
 
     try (CloseableIterable<RowData> reader =
         Parquet.read(Files.localInput(testFile))
             .project(schema)
             .createReaderFunc(type -> FlinkParquetReaders.buildReader(schema, type))
             .build()) {
       Iterator<RowData> rows = reader.iterator();
-      Assert.assertTrue("Should have at least one row", rows.hasNext());
+      assertThat(rows).as("Should have at least one row").hasNext();
+      RowData rowData = rows.next();
+      assertThat(expectedByte).isEqualTo(rowData.getArray(0).getBinary(0));
+      assertThat(expectedByte).isEqualTo(rowData.getBinary(1));
+      assertThat(rows).as("Should not have more than one row").isExhausted();
+    }
+  }
+
+  @Test
+  public void testReadBinaryFieldAsString() throws IOException {
+    Schema schemaForWriteBinary = new Schema(optional(1, "strbytes", Types.BinaryType.get()));
+    org.apache.avro.Schema avroSchema = AvroSchemaUtil.convert(schemaForWriteBinary.asStruct());
+
+    File testFile = temp.newFile();
+    assertThat(testFile.delete()).isTrue();
+
+    String expectedString = "hello";
+
+    GenericRecordBuilder recordBuilder = new GenericRecordBuilder(avroSchema);
+    ByteBuffer expectedBinary = ByteBuffer.wrap(expectedString.getBytes(StandardCharsets.UTF_8));
+    recordBuilder.set("strbytes", expectedBinary);
+    GenericData.Record expectedRecord = recordBuilder.build();
+
+    try (ParquetWriter<GenericRecord> writer =
+        AvroParquetWriter.<GenericRecord>builder(new Path(testFile.toURI()))
+            .withDataModel(GenericData.get())
+            .withSchema(avroSchema)
+            .build()) {
+      writer.write(expectedRecord);
+    }
+
+    // read as string
+    Schema schemaForReadBinaryAsString =
+        new Schema(optional(1, "strbytes", Types.StringType.get()));
+    try (CloseableIterable<RowData> reader =
+        Parquet.read(Files.localInput(testFile))
+            .project(schemaForReadBinaryAsString)
+            .createReaderFunc(
+                type -> FlinkParquetReaders.buildReader(schemaForReadBinaryAsString, type))
+            .build()) {
+      Iterator<RowData> rows = reader.iterator();
+      assertThat(rows).as("Should have at least one row").hasNext();
+      RowData rowData = rows.next();
+      assertThat(rowData.getString(0)).isInstanceOf(BinaryStringData.class);
+      assertThat(rowData.getString(0).toString()).isEqualTo(expectedString);
+      assertThat(rows).as("Should not have more than one row").isExhausted();
+    }
+
+    // read as byte[]
```

Comment:
Should this be a separate test case?

Reply:
This unit test is testing whether expected values can be read from same file and column when the iceberg column definition is both string and binary.